crates/path_trie/src/lib.rs

```diff
@@ -0,0 +1,967 @@
+                    return; // don't descend, file shadows
+                }
+                for (comp, child) in &n.children {
+                    cur.push(comp);
+                    collect_new_winners(child, cur, new_rank, from_clobbers);
+                    cur.pop();
+                }
+            }
+        }
+
+        fn dfs(
+            n: &Node,
+            cur: &mut PathBuf,
+            old_rank: &HashMap<&PackageName, usize>,
+            new_rank: &HashMap<&PackageName, usize>,
+            to_clobbers: &mut ToClobbers,
+            from_clobbers: &mut FromClobbers,
+        ) {
+            let old_winner = n
+                .prefixes
+                .iter()
+                .max_by_key(|p| old_rank.get(p).copied().unwrap_or(usize::MAX));
+            let new_winner = n
+                .prefixes
+                .iter()
+                .max_by_key(|p| new_rank.get(p).copied().unwrap_or(usize::MAX));
+
+            let old_is_file = old_winner.is_some_and(|p| n.terminals.contains(p));
+            let new_is_file = new_winner.is_some_and(|p| n.terminals.contains(p));
+
+            match (old_is_file, new_is_file) {
+                (true, true) => {
+                    let old = old_winner.unwrap();
+                    let new = new_winner.unwrap();
+                    if old != new {
+                        to_clobbers.push((cur.clone(), old.clone()));
+                        from_clobbers.push((cur.clone(), new.clone()));
+                    }
+                    // don't descend
+                }
+                (false, true) => {
+                    let new = new_winner.unwrap();
+                    from_clobbers.push((cur.clone(), new.clone()));
+                    collect_removed_subtree(n, cur, old_rank, to_clobbers);
+                    // don't descend
+                }
+                (true, false) => {
+                    let old = old_winner.unwrap();
+                    to_clobbers.push((cur.clone(), old.clone()));
+                    collect_new_winners(n, cur, new_rank, from_clobbers);
+                    // don't descend further: new_winners handles it
+                }
+                (false, false) => {
+                    // Recurse normally
+                    for (comp, child) in &n.children {
+                        cur.push(comp);
+                        dfs(child, cur, old_rank, new_rank, to_clobbers, from_clobbers);
+                        cur.pop();
+                    }
+                }
+            }
+        }
+
+        {
+            let sorted_packages = self
+                .packages
+                .iter()
+                .cloned()
+                .sorted()
+                .collect::<Vec<PackageName>>();
+
+            let new_sorted_packages = new_order
+                .iter()
+                .cloned()
+                .sorted()
+                .collect::<Vec<PackageName>>();
+
+            assert_eq!(
```

Comment:
Maybe we can turn this into a `debug_assert_eq` to avoid comparing lots of sets?

Reply:
I that we should keep this assertion in release as well to catch potential bugs, at least for some time, then we can change it to debug. Without it we could get hard to debug issues.

I changed implementation to one with linear time and memory, so it should be fast now.